src/graph/sampler.cc

```diff
@@ -0,0 +1,463 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/sampler.cc
+ * \brief DGL sampler implementation
+ */
+
+#include <dgl/sampler.h>
+#include <dgl/immutable_graph.h>
+#include <algorithm>
+
+#ifdef _MSC_VER
+// rand in MS compiler works well in multi-threading.
+int rand_r(unsigned *seed) {
+  return rand();
+}
+#endif
```

Comment:
Can we use std::random from c11 to avoid this platform check?

Reply:
we need to use rand_r for the sake of performance. rand or std::random has a global variable for seed. It causes very bad performance in multithreads.

Reply:
There seems to be solution from [here](https://stackoverflow.com/questions/24007263/is-random-number-generation-in-multiple-threads-using-c11-random-library-slow). I'm actually curious about the exact difference. The similar situation happens in #392 . We could improve this there.